src/FileInputFieldWithList.js

```diff
@@ -0,0 +1,152 @@
+import React, { Component } from 'react'
+import propTypes from '@dhis2/prop-types'
+
+import { FileInputField } from './FileInputField.js'
+import { FileListItemWithRemove } from './FileInputFieldWithList/FileListItemWithRemove.js'
+import { statusPropType, sizePropType } from './common-prop-types.js'
+
+/* 
+NOTE:
+A JavaScript FileList instance is read-only, so we cannot add files to it
+FileList also doesn't have a .map method so by destructuring the FileList
+instance into an array we can add, remove and map
+*/
+
+class FileInputFieldWithList extends Component {
+    handleChange = ({ files: fileList }, event) => {
+        const { onChange, name } = this.props
+
+        onChange({ files: this.updateFileArray(fileList), name: name }, event)
+    }
+
+    updateFileArray(fileList) {
+        const { multiple, files } = this.props
+
+        if (!multiple) {
+            // Spread immutable FileList instance onto array
+            return [...fileList]
+        }
+
+        return [...files, ...fileList].reduceRight(
+            (acc, file) => {
+                if (!acc.unique.has(file.name)) {
+                    acc.unique.add(file.name)
+                    acc.files.unshift(file)
+                }
+                return acc
+            },
+            { files: [], unique: new Set() }
```

Comment:
Is there a reason why you used `reduceRight` and `unshift` instead of `reduce` and `push`?

Reply:
I've looked into this some more and will past a section of our Slack conversation below:

The current setup works, but it is quite important to realise that:
1. The File instances connected to an input is a dynamic reference to a file in the file system, so whenever the file is updated, that update is “propagated to the input”. For example, the `lastModified` date etc are updated.
2. When adding the same file to our FileInput a second time, we create a **new reference** to the same file, i.e. we have two references to the same thing. The references themselves are in-equal but the underlying file is the same.

So what to conclude from that:
1. We only need to make sure we de-duplicate the list. I previously thought I needed to make sure we use the “most recently added” version of the file in case of file-edits etc. But this is not the case.
2. For spotting duplicates we can’t use object equality. But, having said that, checking both the `name` and `lastModified` property on the file should be good enough… (in a later part of the discussion we decided to add `size` too.

I've updated the implementation to reflect the above.

Reply:
To clarify: this method needed some attention anyway, and when doing so, I got rid of the `unshift` and `reduceRight`.